src/connectors/source.rs

```diff
@@ -0,0 +1,1052 @@
+                        error!("[Connector::{}] reader error: {}", ctx.url, e);
+                        break;
+                    }
+                };
+                let last = matches!(&sc_data, SourceReply::EndStream(_));
+                if tx.send(sc_data).await.is_err() || last {
+                    break;
+                };
+            }
+            if reader.on_done(stream) == StreamDone::ConnectorClosed {
+                if let Err(e) = ctx.notifier.notify().await {
+                    error!("[Connector::{}] Failed to notify connector: {}", ctx.url, e);
+                };
+            }
+        });
+    }
+}
+
+#[async_trait::async_trait()]
+impl Source for ChannelSource {
+    async fn pull_data(&mut self, _pull_id: u64, _ctx: &SourceContext) -> Result<SourceReply> {
+        match self.rx.try_recv() {
+            Ok(reply) => Ok(reply),
+            Err(TryRecvError::Empty) => {
+                // TODO: configure pull interval in connector config?
+                Ok(SourceReply::Empty(10))
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// this source is not handling acks/fails
+    fn is_transactional(&self) -> bool {
+        false
+    }
+}
+
+// TODO make fields private and add some nice methods
+/// context for a source
+#[derive(Clone)]
+pub struct SourceContext {
+    /// connector uid
+    pub uid: u64,
+    /// connector url
+    pub url: TremorUrl,
+    /// The Quiescence Beacon
+    pub quiescence_beacon: QuiescenceBeacon,
+}
+
+/// address of a source
+#[derive(Clone, Debug)]
+pub struct SourceAddr {
+    /// the actual address
+    pub addr: Sender<SourceMsg>,
+}
+
+#[allow(clippy::module_name_repetitions)]
+pub struct SourceManagerBuilder {
+    qsize: usize,
+    streams: Streams,
+    source_metrics_reporter: SourceReporter,
+}
+
+impl SourceManagerBuilder {
+    pub fn qsize(&self) -> usize {
+        self.qsize
+    }
+
+    pub fn spawn<S>(self, source: S, ctx: SourceContext) -> Result<SourceAddr>
+    where
+        S: Source + Send + 'static,
+    {
+        let qsize = self.qsize;
+        let name = ctx.url.short_id("c-src"); // connector source
+        let (source_tx, source_rx) = bounded(qsize);
+        let manager = SourceManager::new(source, ctx, self, source_rx);
+        // spawn manager task
+        task::Builder::new().name(name).spawn(manager.run())?;
+
+        Ok(SourceAddr { addr: source_tx })
+    }
+}
+
+/// create a builder for a `SourceManager`.
+/// with the generic information available in the connector
+/// the builder then in a second step takes the source specific information to assemble and spawn the actual `SourceManager`.
+pub fn builder(
+    connector_uid: u64,
+    config: &ConnectorConfig,
+    connector_default_codec: &str,
+    qsize: usize,
+    source_metrics_reporter: SourceReporter,
+) -> Result<SourceManagerBuilder> {
+    let preprocessor_names = config.preprocessors.clone().unwrap_or_else(Vec::new);
```

Comment:
`unwrap_or_default` might be neater here :P

Reply:
Also why not turn this function into `SourceManagerBuilder::new`?

Reply:
It is my past as a java developer which is haunting my code to this day.